packages/block-library/src/page-list/index.php

```diff
@@ -0,0 +1,217 @@
+ *
+ * @param  array $context Navigation block context.
+ * @return array Font size CSS classes and inline styles.
+ */
+function block_core_page_list_build_css_font_sizes( $context ) {
+	// CSS classes.
+	$font_sizes = array(
+		'css_classes'   => array(),
+		'inline_styles' => '',
+	);
+
+	$has_named_font_size  = array_key_exists( 'fontSize', $context );
+	$has_custom_font_size = array_key_exists( 'customFontSize', $context );
+
+	if ( $has_named_font_size ) {
+		// Add the font size class.
+		$font_sizes['css_classes'][] = sprintf( 'has-%s-font-size', $context['fontSize'] );
+	} elseif ( $has_custom_font_size ) {
+		// Add the custom font size inline style.
+		$font_sizes['inline_styles'] = sprintf( 'font-size: %spx;', $context['customFontSize'] );
+	}
+
+	return $font_sizes;
+}
+
+/**
+ * Outputs Page list markup from an array of pages with nested children.
+ *
+ * @param array $nested_pages The array of nested pages.
+ *
+ * @return string List markup.
+ */
+function render_nested_page_list( $nested_pages ) {
+	if ( empty( $nested_pages ) ) {
+		return;
+	}
+	$markup = '';
+	foreach ( (array) $nested_pages as $page ) {
+		$css_class = 'wp-block-pages-list__item';
+		if ( isset( $page['children'] ) ) {
+			$css_class .= ' has-child';
+		}
+		$markup .= '<li class="' . $css_class . '">';
+		$markup .= '<a class="wp-block-pages-list__item__link" href="' . esc_url( $page['link'] ) . '">' . wp_kses(
+			$page['title'],
+			wp_kses_allowed_html( 'post' )
+		) . '</a>';
+		if ( isset( $page['children'] ) ) {
+			$markup .= '<span class="wp-block-page-list__submenu-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" viewBox="0 0 24 24" transform="rotate(90)"><path d="M8 5v14l11-7z"></path><path d="M0 0h24v24H0z" fill="none"></path></svg></span>';
+			$markup .= '<ul class="submenu-container">' . render_nested_page_list( $page['children'] ) . '</ul>';
+		}
+		$markup .= '</li>';
+	}
+	return $markup;
+}
+
+/**
+ * Outputs nested array of pages
+ *
+ * @param array $current_level The level being iterated through.
+ * @param array $children The children grouped by parent post ID.
+ *
+ * @return array The nested array of pages.
+ */
+function nest_pages( $current_level, $children ) {
+	if ( empty( $current_level ) ) {
+		return;
+	}
+	foreach ( (array) $current_level as $key => $current ) {
+		if ( isset( $children[ $key ] ) ) {
+			$current_level[ $key ]['children'] = nest_pages( $children[ $key ], $children );
+		}
+	}
+	return $current_level;
+}
+
+/**
+ * Renders the `core/page-list` block on server.
+ *
+ * @param array $attributes The block attributes.
+ * @param array $content The saved content.
+ * @param array $block The parsed block.
+ *
+ * @return string Returns the page list markup.
+ */
+function render_block_core_page_list( $attributes, $content, $block ) {
+	static $block_id = 0;
+	$block_id++;
+
+	$all_pages = get_pages( array( 'sort_column' => 'menu_order' ) );
```

Comment:
@tellthemachines: Is there a discussion somewhere in GH about limiting this query with `number`?

Reply:
I don't recall there being any discussion on that. I did test it with lots (several hundreds) of pages and it works ok because the block is server rendered. If we wanted to render it on the client side in the editor (I recall some talk about this at one point - possibly with @gwwar ?) it might be a different matter.

Reply:
I think some reasonable limit could be set for both server/client. Any client side implementation will hit issues much sooner (eg needing to fetch all results via API+ potentially adding a ton of navigation links). 

There's should already be a limit for the page-list -> navigation links conversion, where we won't display the option if there are too many pages.